apc_cache.c

```diff
@@ -898,6 +898,53 @@ PHP_APCU_API zend_bool apc_cache_update(
 }
 /* }}} */
 
+/* {{{ apc_cache_atomic_update_long */
+PHP_APCU_API zend_bool apc_cache_atomic_update_long(
+		apc_cache_t *cache, zend_string *key, apc_cache_atomic_updater_t updater, void *data,
+		zend_bool insert_if_not_found, zend_long ttl)
+{
+	apc_cache_entry_t *entry;
+	zend_bool retval = 0;
+	time_t t = apc_time();
+
+	if (!cache) {
+		return 0;
+	}
+
+retry_update:
+	APC_RLOCK(cache->header);
+	entry = apc_cache_rlocked_find_nostat(cache, key, t);
```

Comment:
As the updater is quick and there is just one ATOMIC operation, it is not faster to get an incref lock on the entry itself. (as that would be 3 atomic operations).

But I wondered:

Would it not be enough to have an incref on the entry to be able to update it in general?

---

I'll do some benchmarks with this PR :)

Reply:
The incref part is what allows you to use the entry after releasing the rlock -- but this only makes sense if you expect to perform some expensive operation on it. The updater is cheap enough to call it while still holding the rlock.

Acquiring the rlock is necessary to avoid concurrent modifications while the list of entries is traversed. Supporting that without a lock would require a lock-free table.

Reply:
Yep, that is what I meant: If in general instead of insert + delete for apcu_store() it could be enough to do an update with increased incref.